pkg/cachemanager/cachemanager.go

```diff
@@ -0,0 +1,385 @@
+		return
+	}
+
+	c.processExcluder.Replace(newExcluder)
+	// there is a new excluder which means we need to schedule a wipe for any
+	// previously watched GVKs to be re-added to get a chance to be evaluated
+	// for this new process excluder.
+	c.excluderChanged = true
+}
+
+func (c *CacheManager) watchesGVK(gvk schema.GroupVersionKind) bool {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	return c.watchedSet.Contains(gvk)
+}
+
+func (c *CacheManager) AddObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	gvk := instance.GroupVersionKind()
+
+	isNamespaceExcluded, err := c.processExcluder.IsNamespaceExcluded(process.Sync, instance)
+	if err != nil {
+		return fmt.Errorf("error while excluding namespaces for gvk: %+v: %w", gvk, err)
+	}
+
+	if isNamespaceExcluded {
+		c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+		return nil
+	}
+
+	syncKey := syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName())
+	if c.watchesGVK(gvk) {
+		_, err = c.cfClient.AddData(ctx, instance)
+		if err != nil {
+			c.syncMetricsCache.AddObject(
+				syncKey,
+				syncutil.Tags{
+					Kind:   instance.GetKind(),
+					Status: metrics.ErrorStatus,
+				},
+			)
+
+			return err
+		}
+
+		c.syncMetricsCache.AddObject(syncKey, syncutil.Tags{
+			Kind:   instance.GetKind(),
+			Status: metrics.ActiveStatus,
+		})
+		c.syncMetricsCache.AddKind(instance.GetKind())
+	}
+
+	c.tracker.ForData(instance.GroupVersionKind()).Observe(instance)
+
+	return nil
+}
+
+func (c *CacheManager) RemoveObject(ctx context.Context, instance *unstructured.Unstructured) error {
+	gvk := instance.GroupVersionKind()
+
+	if c.watchesGVK(gvk) {
+		if _, err := c.cfClient.RemoveData(ctx, instance); err != nil {
+			return err
+		}
+	}
+
+	// only delete from metrics map if the data removal was successful
+	c.syncMetricsCache.DeleteObject(syncutil.GetKeyForSyncMetrics(instance.GetNamespace(), instance.GetName()))
+	c.tracker.ForData(instance.GroupVersionKind()).CancelExpect(instance)
+
+	return nil
+}
+
+func (c *CacheManager) wipeData(ctx context.Context) error {
+	if _, err := c.cfClient.RemoveData(ctx, target.WipeData()); err != nil {
+		return err
+	}
+
+	// reset sync cache before sending the metric
+	c.syncMetricsCache.ResetCache()
+	c.syncMetricsCache.ReportSync()
+
+	return nil
+}
+
+func (c *CacheManager) ReportSyncMetrics() {
+	c.syncMetricsCache.ReportSync()
+}
+
+func (c *CacheManager) syncGVK(ctx context.Context, gvk schema.GroupVersionKind) error {
+	u := &unstructured.UnstructuredList{}
+	u.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   gvk.Group,
+		Version: gvk.Version,
+		Kind:    gvk.Kind + "List",
+	})
+
+	var err error
+	func() {
+		c.mu.Lock()
+		defer c.mu.Unlock()
+
+		// only call List if we are still watching the gvk.
+		if c.watchedSet.Contains(gvk) {
+			err = c.reader.List(ctx, u)
+		}
+	}()
+
+	if err != nil {
+		return fmt.Errorf("replaying data for %+v: %w", gvk, err)
```

Comment:
nit: s/replaying data for/listing/g

This method doesn't really know what replaying is.  It just knows how to list something and iterate over it.